libpromises/evalfunction.c

```diff
@@ -8450,6 +8451,191 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
+                                       const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+
+    struct ThreadInfo thread_info = {0};
+
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires path as first argument",
+            fp->name);
+        return FnFailure();
+    }
+    thread_info.path = RlistScalarValue(finalargs);
+
+    long timeout = (finalargs->next == NULL) ? 0L
+                 : IntFromString(RlistScalarValue(finalargs->next));
+    assert(timeout >= 0);
+
+    if (timeout == 0)
+    {
+        char buf[1];
+        const int fd = open(thread_info.path, O_RDONLY | O_NONBLOCK);
+        if (fd < 0) {
+            Log(LOG_LEVEL_DEBUG, "Failed to open file '%s': %s",
+                thread_info.path, GetErrorStr());
+            return FnReturnContext(false);
+        }
+        else if (read(fd, buf, sizeof(buf)) < 0)
+        {
+            Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s",
+                thread_info.path, GetErrorStr());
+            close(fd);
+            return FnReturnContext(false);
+        }
+        close(fd);
+        return FnReturnContext(true);
+    }
+
+    struct timespec timespec = {0};
+    xclock_gettime(CLOCK_REALTIME, &timespec);
+    timespec.tv_sec += timeout;
+
+    int ret = pthread_mutex_init(&(thread_info.mutex), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_mutex_lock(&(thread_info.mutex));
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_create(&(thread_info.thread), NULL, &ThreadRoutine, &thread_info);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create thread: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_init(&(thread_info.cond), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize condition: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_timedwait(&(thread_info.cond), &(thread_info.mutex),
+                                 &timespec);
+    switch (ret)
+    {
+    case 0:
+        {
+            ret = pthread_join(thread_info.thread, NULL);
+            if (ret != 0)
+            {
+                Log(LOG_LEVEL_ERR, "Failed to join thread: %s",
+                    GetErrorStrFromCode(ret));
+                return FnFailure();
+            }
+            return thread_info.result;
+        }
+
+    case ETIMEDOUT:
+        {
+            Log(LOG_LEVEL_DEBUG, "File '%s' is not readable: "
+                "Read operation timed out, exceeded %ld seconds.",
+                thread_info.path, timeout);
+
+            ret = pthread_detach(thread_info.thread);
```

Comment:
maybe `pthread_kill()` would not hurt here as well

Reply:
> But note  also  that POSIX  says that an attempt to use a thread ID whose lifetime has ended produces undefined behavior, and an attempt to use an invalid thread ID in  a  call  to  pthread_kill()  can, for example, cause a segmentation fault.

This sounds kind of dangerous. Unless the thread is still considered alive until you call `pthread_join`? Maybe we can use `pthread_cancel` instead?

Reply:
> Unless the thread is still considered alive until you call pthread_join? 

Yes, for a joinable thread, this should be the case. So `pthread_kill()` followed by `pthread_detach()` ought to be OK. :wink: `pthread_cancel()` is an alternative, but AFAICT `pthread_kill()` has the benefit of interrupting the potential hanging syscall (`open()`,...). But again, exotics... I cannot even 100% say that `pthread_kill()` won't kill the whole process there. :grin:

Reply:
Used `pthread_cancel` alternative